Point turn rebased

AG6GR opened the pull request:
With the power of rebase -i and cherry-pick, I've reconstructed the PointTurn branch on top of master branch. Diff reports no difference from the original PointTurn branch besides the new gitignore, no more .DS_Store files, and no dist/FRCUserProgram.jar

Please verify that this builds correctly on your setup, and use this branch for PointTurn code review.

Commit:
initial point turn commit

mvoodarla committed:
Fixed issue with slow check being negative

mvoodarla committed:
point turn tested but still innacurate by a factor for every angle

Commit:
removed unecessary classes and renamed drive straight

Commit:
changed PointTurn execute to dampen angle after a degree threshold

tnarayanan committed:
Fixed gradual slowdown, need to fix overshoot after robot stop

iamef committed:
modified linearization of power calibration, need to test

mvoodarla committed:
added further dampening to point turn at angle closer to goal

tnarayanan committed:
Merge pull request #3 from Tino-FRC-2473/BasicDriveCode

basic drive code functionality

mvoodarla committed:
Stall experiment changes, K_TURN change to PointTurn

mvoodarla committed:
Working 90 degree turn with negative slowdown

mvoodarla committed:
update gitignore

tnarayanan committed:
Successful command group testing with point turn and drive straight working pretty well

mvoodarla committed:
Reset encoders in auto init

Earlier, we were resetting encoders in robot init which caused resetting
problems in drive straight but now it is fixed.

mvoodarla committed:
Temporary commit

My battery is low

Commit:
Ensure sign of encoder ticks is consistent

Before, our front right encoders were reporting negative even
though the robot was moving forward. This caused error in drive
straight and turning. Change getEncoderTicks method to reverse
sign of right side.

tnarayanan committed:
Fix StraightDrive for negative distances

StraightDrive has been updated to accomodate driving backwards.
Additionally, a temporary fix for encoder counts has been added to wait
until the encoder value reaches zero before resetting the next encoder.
This was added because encoder values were still getting reset in the
middle of the StraightDrive command, despite a delay of 5000ms in the
reset encoder function.

michelllepan committed:
Change encoder resets to asynchronous

Previously we used a 5000 ms timeout in setSelectedSensorPosition(), but
we removed this timeout and made our resetEncoders() method synchronized
so that the robot intialization will wait for the encoders to reset
before moving on.

AG6GR commented on src/org/usfirst/frc/team2473/robot/subsystems/DriveSubsystem.java:
```diff
@@ -150,27 +154,14 @@ public int getEncoderTicks(int id) {
-		}
-		
-		while (motor.getSelectedSensorPosition() != 0);
+	public synchronized void resetEncoders() {
```
The synchronized keyword means this won't be called at the same time as other code, but do you actually wait for the reset to finish?

michelllepan replied:
resolved by 65cb5ab

AG6GR commented on src/org/usfirst/frc/team2473/robot/Robot.java:
```diff
@@ -7,114 +7,110 @@
-			m_autonomousCommand.start();
-		}
+		driveSubsystem.resetEncoders();
+//		double speed = prefs.getDouble("speed", 0.4);
```
Are these commented lines still needed?

AG6GR commented on src/org/usfirst/frc/team2473/robot/RobotMap.java:
```diff
@@ -7,20 +7,21 @@
-	// number and the module. For example you with a rangefinder:
-	// public static int rangefinderPort = 1;
-	// public static int rangefinderModule = 1;
+	public static int TALON_BL = 3;
```
Can these constants be marked final?

AG6GR commented on src/org/usfirst/frc/team2473/robot/commands/AutonomousTester.java:
```diff
@@ -0,0 +1,46 @@
+package org.usfirst.frc.team2473.robot.commands;
```
There's some real funky whitespace going on here. Let's indent with tabs or spaces, but not both.

AG6GR replied:
Resolved by f676c45

AG6GR commented on src/org/usfirst/frc/team2473/robot/commands/AutonomousTester.java:
```diff
@@ -0,0 +1,46 @@
+	      addSequential(new StraightDrive(distanceSecond, 0.3));
+        
+    	
+//    	for (int i = 0; i < 4; i++) {
```
Comment what each of these options are. To be super clean, try using a switch and a hardcoded constant to select a specific pattern.

AG6GR replied:
Resolved in f676c45. Comments are great btw.

AG6GR commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -0,0 +1,126 @@
+	private double initialAngle;
+	private double initialPower;	
+	
+	public PointTurn(double degrees, double power) {
```
"degrees" and "power" is a bit vague, maybe targetDegrees? Also a comment to be sure.

AG6GR commented on src/org/usfirst/frc/team2473/robot/commands/PointTurn.java:
```diff
@@ -0,0 +1,126 @@
+	}
+	
+	private void setPower(double power) {
+		if (power < 0) throw new IllegalArgumentException("Power must be a positive scalar for point turn!");
```
I'll let the lack of braces slide, but do put the body of the if a new line. Otherwise it can be quite hard to read, especially if you also mix in ternary operators.

AG6GR commented on src/org/usfirst/frc/team2473/robot/commands/StallExperiment.java:
```diff
@@ -0,0 +1,75 @@
+                                  RobotMap.TALON_FL, RobotMap.TALON_BL};
+    
+
+    public StallExperiment(double power, double change) {
```
What is change?

michelllepan replied:
```change``` is the amount the power is decreased by each time – we'll change the name to ```decrementBy``` to make this more clear.

mvoodarla replied:
Ooh I changed it to `powerDecrement`

AG6GR commented on src/org/usfirst/frc/team2473/robot/subsystems/DriveSubsystem.java:
```diff
@@ -0,0 +1,176 @@
+	
+	DifferentialDrive teleopDrive;
+	
+	private HashMap<Double, Double> leftTable = new HashMap<>();
```
Might be good to write down what the inputs and outputs of these lookup tables are.

AG6GR commented on src/org/usfirst/frc/team2473/robot/subsystems/DriveSubsystem.java:
```diff
@@ -0,0 +1,176 @@
+	public double convertPower(double power, boolean isLeft) {
+		HashMap<Double, Double> tempTable = isLeft ? leftTable : rightTable;
+		if (power < minTestedPower) {
+			return power * 1/tempTable.get(minTestedPower);
```
Is the 1 necessary? And is it the right type?

mvoodarla replied:
We use the 1 so that the powers generated will always be less than the power that has been inputted. We thought this was reasonable since it's probably unsafe to move motors at a power that are higher than specified. Also, I think the type of 1 is correct since all the other values used are doubles.

michelllepan replied:
I think what Sunny meant is that we can just divide by ```tempTable.get(minTestedPower)``` instead of multiplying by 1 and then dividing.

AG6GR commented on src/org/usfirst/frc/team2473/robot/subsystems/DriveSubsystem.java:
```diff
@@ -0,0 +1,176 @@
+			}
+			
+			int i = 0;
+			while(powers.get(i) < power) i++;
```
I challenge you to rewrite this as a one line for loop. Also what is up with that `i--`?

michelllepan replied:
We want to find the greatest ```i``` for which ```powers.get(i)``` is less than ```power```, which requires decrementing once after the while loop, but we can loop backwards to make this cleaner.

michelllepan replied:
Also this wasn't written as a for loop because we need to access ```i``` afterwards.

AG6GR replied:
You can declare int i; ahead of the loop so it is within the entire function's scope.

mvoodarla committed:
Remove unused constants and make talon id's final

This will clean up the RobotMap.java file so that we only have the
constants that we're using.

mvoodarla committed:
Remove unused Testing.java file

mvoodarla committed:
Make PointTurn.java cleaner

Changed parameter names and removed a few comments

michelllepan committed:
Fix whitespace in AutonomousTester.java

Indentation is now only with tabs instead of using both spaces and tabs.

michelllepan committed:
Clean up AutonomousTester.java and add comments

Each sequence (square, lightning bolt, etc.) is now in a separate method
that can be called from Robot.java. The constructor has also been modified
to include a parameter for wait time, and autonomousInit() in Robot.java
has been changed accordingly.

michelllepan committed:
Change wait variable in AutonomousTester to a constant

mvoodarla committed:
Clean up StallExperiment with variable name changes and import removals

Changed a parameter variable name from change to powerDecrement. Also,
removed unused import statements.

mvoodarla committed:
Include documentation for PointTurn

Completely commented the PointTurn class.

michelllepan committed:
Fix encoder resets in DriveSubsystem and add documentation

All methods involving encoder values are now synchronized so that
encoder values cannot be retrieved while the encoders are being reset.

michelllepan committed:
Merge branch 'PointTurn_Rebased' of https://github.com/Tino-FRC-2473/FRC2018-Offseason into PointTurn_Rebased

mvoodarla committed:
Removed uneccesary comments from Robot.java

mvoodarla committed:
Add documentation to StallExperiment.java

mvoodarla committed:
Add documentation to TeleopDrive.java

mvoodarla committed:
Add documentation to StraightDrive.java

Added comments to this class and changed the names of a few variables
for cleanliness.

mvoodarla reviewed (approved):
Looks good now that all comments have been resolved.

tnarayanan committed:
Add documentation to Robot.java

Added comments to the different methods in Robot.java and removed
unnecessary imports.
